add an auditd container

tych0 opened the pull request:
Two things to note here: we unfortunately can't just exec audit, because it
dies if /var/log/audit doesn't exist, so we have to do a wrapper script.

Second, it also dies if it can't re-nice itself, so we have to give it
CAP_SYS_NICE as well as the audit caps.

I didn't add this to the default linuxkit.yml because the linuxkit/audit
repo doesn't exist yet, but we should probably (?) do that.

Closes #52

Signed-off-by: Tycho Andersen <tycho@docker.com>

justincormack commented:
Is there a way to make it log to `stdout` instead? We are going to be reworking logs for remote logging etc, and we can't really have programs owning their log files, or we can't do rotation etc.

tych0 commented:
Yep, it's a pretty easy change, I can make it now.

ijc commented:
> /var/log/audit doesn't exist, so we have to do a wrapper script.

You can create directories in the `files` section of the yaml, but I guess that might be tricky to rely on and a bit of learning-curve/pain for every user who forgets/doesn't know. Perhaps the label stuff could be extended to allow making dirs?

Or if you make the binds `/var/log:/var/log` *and* `/var/log/auditd:/var/log/auditd` will it be auto created? I can't remember sadly.

There's also the option of a wishlist bug upstream to add a `mkdir`...

justincormack commented:
No, we can't create the directories when the bind is specified as the filesystem is read only. We could add directories that there are binds for at build time I guess.

tych0 commented:
Yep, and if we run it logging to stdout, we don't even need that dir, so I removed that bit.

Also note that events like USER_AUTH and USER_LOGIN are recorded from userspace (via pam or the actual thing itself), and busybox doesn't seem to send these signals, but the "real" versions of them do. So if you spend a few hours trying to figure out why the USER_ events don't seem to be logged, that's why :).

Anyway, above is the updated patch, with logging to stdout and the same set of default rules as ubuntu uses.

tych0 commented:
Hmm. SIGBUS? Seems like the ci might be angry?

justincormack commented on pkg/auditd/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+COPY audit.rules /etc/audit
+COPY runaudit.sh /usr/bin
+
+CMD ["/sbin/tini", "/usr/bin/runaudit.sh"]
```
I don't think we would need `tini` as I would not expect `auditd` to even fork?

tych0 replied:
It depends on how you have it configured. If you use a dispatcher with it, it forks the dispatcher and that itself forks whatever you tell it to. IIRC it also forks when parsing config for some reason.

justincormack replied:
ah ok.

justincormack commented:
I re-ran the CI as it seemed to hang right at the end shutting down GCP this time

justincormack commented:
The efi ISO test is failing, possibly for unrelated reasons... but it is consistent.

justincormack commented:
Can you rebase, CI should be fixed after.

riyazdf commented on pkg/auditd/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+# FROM linuxkit/alpine:630ee558e4869672fae230c78364e367b8ea67a9 AS mirror
```
should we just remove this line?  Do we know when mainline alpine might get the package?

tych0 replied:
I have no idea what the promotion process is, unfortunately. I suppose we could talk to @ncopa about it :)

tych0 replied:
Anyway, yes, sorry about that, I dropped the line.

riyazdf commented on pkg/auditd/Makefile:
```diff
@@ -0,0 +1,15 @@
+.PHONY: tag push
```
can we use the new style Makefile?  Should be pretty simple, see https://github.com/linuxkit/linuxkit/blob/master/pkg/rngd/Makefile as an example

tych0 replied:
Yep, done.

justincormack commented on pkg/auditd/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+FROM alpine:edge AS mirror
```
This is not reproducible, so we can't use this like this with tree hashes. The options are either to pin this to a specific sha256 hash, or to build auditd from source, or not use the standard Makefile.

riyazdf replied:
ah I missed this - good point.  Let's use `alpine:edge@sha256:b45de39913783fcc1ee31e15bc9a3df4226a0a51111e00ca74ea8531c099d4df`

justincormack replied:
That is still not reproducible as it installs packages.

tych0 replied:
Building from source won't be reproducible either.

I don't see a way to do a reproducible build without essentially replicating what we have for main for this. In any case, per a talk with @ncopa I've started the discussion about main inclusion for audit: https://github.com/alpinelinux/aports/pull/1776

justincormack replied:
building from source is reproducible if you check out an exact git commit or downloaded version, thats what we do elsewhere.

ijc replied:
I think we don't get exact binary reproducibility with this method, unless the code base happens to support it, in general we only get the same functional output given the same inputs i.e. without variation in components due to network mirrors etc changing. (I suspect you are each talking about a different level of "reproducible" here).

justincormack replied:
Sure, if it embeds build timestamps or whatever, but the requirement at present is that all the inputs are covered by the tree hash. We will work on the other parts later.

tych0 replied:
Ok, I set it up to build from source. We need a few of audits deps (and abuild) in the base image, so there's a commit that adds those. I think CI will fail until that's merged; I'll PR it separately assuming this approach works for people.

justincormack commented:
I think the base alpine package has been updated again since this, so the hash is probably different?

tych0 commented:
I'm including some of the auditd build deps in the base image (added in the first commit in the series), so the base image hash will always be different. That's why I was thinking ci would fail, because it tries to build everything in pkg/.

In any case, you're right that the underlying package list has changed, so I rebased and updated the hash so that it should be right.
